web/packages/teleport/src/UnifiedResources/Resources.tsx

```diff
@@ -50,56 +53,39 @@ export function Resources() {
     fieldName: 'name',
     dir: 'ASC',
   });
-  const {
-    params,
-    search,
-    setParams,
-    replaceHistory,
-    pathname,
-    setSort,
-    onLabelClick,
-  } = filtering;
-
-  const { fetchInitial, fetchedData, attempt, fetchMore } = useInfiniteScroll({
+  const { params, setParams, replaceHistory, pathname, setSort, onLabelClick } =
+    filtering;
+
+  const { fetch, forceFetch, resources, attempt } = useKeyBasedPagination({
     fetchFunc: teleCtx.resourceService.fetchUnifiedResources,
     clusterId,
-    params,
+    filter: params,
   });
 
-  useEffect(() => {
-    fetchInitial();
-  }, [clusterId, search]);
-
-  const infiniteScrollDetector = useRef(null);
-
-  // Install the infinite scroll intersection observer.
-  //
-  // TODO(bl-nero): There's a known issue here. We need to have `fetchMore` in
-  // the list of hook dependencies, because using a stale `fetchMore` closure
-  // means we will fetch the same data over and over. However, as it's
-  // implemented now, every time `fetchMore` changes, we reinstall the observer.
-  // This is mitigated by `fetchMore` implementation, which doesn't spawn
-  // another request before the first one finishes, but it's still a potential
-  // for trouble in future. We need to decouple updating the `fetchMore` closure
-  // and installing the observer.
-  useEffect(() => {
-    if (infiniteScrollDetector.current) {
-      const observer = new IntersectionObserver(entries => {
-        if (entries[0]?.isIntersecting) {
-          fetchMore();
-        }
-      });
-      observer.observe(infiniteScrollDetector.current);
-      return () => observer.disconnect();
-    }
-  });
+  const scrollDetector = useRef(null);
+
+  useInfiniteScroll(scrollDetector.current, fetch);
 
   if (!enabled) {
     history.replace(cfg.getNodesRoute(clusterId));
   }
 
+  const retryClicked = () => {
+    forceFetch();
+  };
+
   return (
     <FeatureBox>
+      {attempt.status === 'failed' && (
+        <ErrorBox>
+          <ErrorBoxInternal>
+            <Danger>
+              {attempt.statusText}
+              <ButtonLink onClick={retryClicked}>Retry</ButtonLink>
```

Comment:
nit: the button hover state very slightly covers up the error message on its left. maybe add a suuuuper small amount of margin left here?
![Screenshot 2023-08-24 at 1 51 05 PM](https://github.com/gravitational/teleport/assets/5201977/95f4da80-d627-4037-a48c-241ead8d5a53)

Reply:
Also, on smaller screensizes we get some weird styles
![Screenshot 2023-08-24 at 1 54 21 PM](https://github.com/gravitational/teleport/assets/5201977/3a18a29b-014d-49f4-a6b3-d3eb665187fb)